src/main/java/org/opensearch/security/resources/ResourceSharingIndexListener.java

```diff
@@ -0,0 +1,130 @@
+     * This method is called during the plugin's initialization process.
+     *
+     * @param threadPool The ThreadPool instance to be used for executing operations.
+     * @param client     The Client instance to be used for interacting with OpenSearch.
+     */
+    public void initialize(ThreadPool threadPool, Client client) {
+
+        if (initialized) {
+            return;
+        }
+
+        initialized = true;
+        this.threadPool = threadPool;
+        this.resourceSharingIndexHandler = new ResourceSharingIndexHandler(
+            ConfigConstants.OPENSEARCH_RESOURCE_SHARING_INDEX,
+            client,
+            threadPool
+        );
+
+    }
+
+    public boolean isInitialized() {
+        return initialized;
+    }
+
+    /**
+     * This method is called after an index operation is performed.
+     * It creates a resource sharing entry in the dedicated resource sharing index.
+     * @param shardId The shard ID of the index where the operation was performed.
+     * @param index The index where the operation was performed.
+     * @param result The result of the index operation.
+     */
+    @Override
+    public void postIndex(ShardId shardId, Engine.Index index, Engine.IndexResult result) {
+
+        String resourceIndex = shardId.getIndexName();
+        log.info("postIndex called on {}", resourceIndex);
+
+        String resourceId = index.id();
+
+        User user = threadPool.getThreadContext().getPersistent(ConfigConstants.OPENDISTRO_SECURITY_USER);
+
+        try {
+            ResourceSharing sharing = this.resourceSharingIndexHandler.indexResourceSharing(
+                resourceId,
+                resourceIndex,
+                new CreatedBy(user.getName()),
+                null
+            );
+            log.info("Successfully created a resource sharing entry {}", sharing);
+        } catch (IOException e) {
+            log.info("Failed to create a resource sharing entry for resource: {}", resourceId);
+        }
+    }
+
+    /**
+     * This method is called after a delete operation is performed.
+     * It deletes the corresponding resource sharing entry from the dedicated resource sharing index.
+     * @param shardId The shard ID of the index where the delete operation was performed.
+     * @param delete The delete operation that was performed.
+     * @param result The result of the delete operation.
+     */
+    @Override
+    public void postDelete(ShardId shardId, Engine.Delete delete, Engine.DeleteResult result) {
+
+        String resourceIndex = shardId.getIndexName();
+        log.info("postDelete called on {}", resourceIndex);
+
+        String resourceId = delete.id();
+
+        boolean success = this.resourceSharingIndexHandler.deleteResourceSharingRecord(resourceId, resourceIndex);
+        if (success) {
+            log.info("Successfully deleted resource sharing entries for resource {}", resourceId);
```

Comment:
Nit. Should this be singular since there would only be one resource sharing entry for each resource?

Reply:
addressed.